lib/puppet/agent.rb

```diff
@@ -29,10 +30,6 @@ def needing_restart?
 
   # Perform a run with our client.
   def run(client_options = {})
-    if running?
-      Puppet.notice "Run of #{client_class} already in progress; skipping  (#{lockfile_path} exists)"
-      return
-    end
```

Comment:
I think we can also remove the call to [`agent.running?`](https://github.com/puppetlabs/puppet/blob/4.0.0/lib/puppet/daemon.rb#L90-L93), and instead have it do:

```
def reload
  return unless agent
  agent.run({:splay => false})
rescue Puppet::LockError
  Puppet.notice "Not triggering already-running agent"
end
```

I'm thinking we should also deprecate the [`Puppet::Agent::Locker#locked?`](https://github.com/puppetlabs/puppet/blob/4.0.0/lib/puppet/util/lockfile.rb#L45-L48) method as it is inherently unsafe. The only safe way to know if the lock is locked and perform some action based on the result, is to try to lock it, and then handle the LockError that may result.

Puppet::Util::Lockfile has similar problems, but I think there's already a ticket about it containing race conditions.